fsmonitor: use an opaque token for the last update

kewillford opened the pull request:
Applications that are used for monitoring the file system might not use a timestamp for querying.  Watchman for example uses a clock id to remove race conditions that come from using a timestamp.

Switch to using an opaque token in the fsmonitor code so that the clock id from watchman or any other token from a file monitor can be stored.

Update the watchman scripts to get the clock id and send that back to the caller.  This is sent back as the first NUL terminated entry in the output stream.  

Still want to clean up the watchman scripts because it is not very well organized or understandable for me at least.

kewillford committed:
fsmonitor: change last update timestamp on the index_state to opaque token

Some file system monitors might not use or take a timestamp for processing
and in the case of watchman could have race conditions with using a
timestamp.

Change the fsmonitor_last_update from being a uint64_t to a char pointer
so that any arbitrary data can be stored in it.

Signed-off-by: Kevin Willford <Kevin.Willford@microsoft.com>

kewillford commented on fsmonitor.c:
```diff
@@ -32,17 +33,28 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
+		strbuf_addstr(&last_update, index);
+		index += last_update.len;
+	} else {
 		return error("bad fsmonitor version %d", hdr_version);
```
Need `stbuf_release` here.

kewillford commented on fsmonitor.c:
```diff
@@ -161,7 +172,7 @@ void refresh_fsmonitor(struct index_state *istate)
 	 * should be inclusive to ensure we don't miss potential changes.
 	 */
-	last_update = getnanotime();
+	query_start = getnanotime();
```
Move this inside the `if` statement below.

kewillford commented on fsmonitor.c:
```diff
@@ -161,7 +172,7 @@ void refresh_fsmonitor(struct index_state *istate)
 	 * This could be racy so save the date/time now and query_fsmonitor
 	 * should be inclusive to ensure we don't miss potential changes.
 	 */
```
Update or delete this comment.

kewillford commented on fsmonitor.c:
```diff
@@ -161,7 +172,7 @@ void refresh_fsmonitor(struct index_state *istate)
+	query_start = getnanotime();
 
 	/*
 	 * If we have a last update time, call query_fsmonitor for the set of
```
Update comment

kewillford commented on t/t7519/fsmonitor-all:
```diff
@@ -1,24 +1,21 @@
-# The hook is passed a version (currently 1) and a time in nanoseconds
+# The hook is passed a version (currently 2) and since token
 # formatted as a string and outputs to stdout all files that have been
 # modified since the given time. Paths must be relative to the root of
```
Update this comment.

kewillford commented on t/t7519/fsmonitor-all:
```diff
@@ -1,24 +1,21 @@
 # modified since the given time. Paths must be relative to the root of
 # the working tree and separated by a single NUL.
 #
 #echo "$0 $*" >&2
```
Remove this?

kewillford commented on fsmonitor.c:
```diff
@@ -6,8 +6,9 @@
-#define HOOK_INTERFACE_VERSION	(1)
+#define INDEX_EXTENSION_VERSION1	(1)
+#define INDEX_EXTENSION_VERSION2	(2)
+#define HOOK_INTERFACE_VERSION		(2)
```
Line this up.

kewillford replied:
This is lined up in my editor so not sure what is going on here.

kewillford commented on templates/hooks--fsmonitor-watchman.sample:
```diff
@@ -3,107 +3,141 @@
-} else {
-	require Cwd;
-	$git_work_tree = Cwd::cwd();
-}
```
@dscho I wasn't sure if the `Cwd` would always be the root of the repo or might be somewhere else in the repository that is why and added the `get_working_dir()` to make sure we are in the root before we add the directory to the watch list.

dscho replied:
I _think_ that the hooks are always called from the top-level directory. If in doubt, maybe there is an environment variable `GIT_WORK_TREE` that you can use?

derrickstolee replied:
This [was discussed recently on-list](https://lore.kernel.org/git/pull.466.git.1574374826.gitgitgadget@gmail.com/T/#t). It looks like `GIT_DIR` is only set if it differs from normal, but maybe `GIT_WORK_TREE` has similar characteristics? At the least, `t3030-merge-recursive.sh` has some `GIT_WORK_TREE` strangeness that can help.

kewillford commented on fsmonitor.c:
```diff
@@ -32,17 +33,28 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 	hdr_version = get_be32(index);
 	index += sizeof(uint32_t);
-	if (hdr_version != INDEX_EXTENSION_VERSION)
+	if (hdr_version == INDEX_EXTENSION_VERSION1) {
```
@dscho Question for you on backwards compatibility.  When there is a new version of a hook like here, what is the usual way of handling that with git? Here I am forcing the version of the hook to be 2, but should I instead be doing something like adding a hook version to the config that defaults to 1 so that version of the hook will continue to work when users install a new version of git?  With it currently anyone with version 1 of the hook will fail until they have a version 2.

dscho replied:
> When there is a new version of a hook like here, what is the usual way of handling that with git?

I don't think that we have a precedent for that yet ;-)

> Here I am forcing the version of the hook to be 2, but should I instead be doing something like adding a hook version to the config that defaults to 1 so that version of the hook will continue to work when users install a new version of git?

I think we'll want to have a config variable, yes, if only to keep regression tests that make sure that Git can still talk to v1 fsmonitor hooks.

> With it currently anyone with version 1 of the hook will fail until they have a version 2.

How about doing a fall-back? Like, defaulting to `-1` for the version, overridable via the config, and if it is `-1`, try 2 first, and if that fails, try 1, and if that succeeds, set the global variable to 1, otherwise to 0 (and consistently skip the fsmonitor for the rest of the process' life time).

What do you think?

kewillford replied:
>How about doing a fall-back?

That would be the most robust.  Thanks for the suggestion.

kewillford committed:
fsmonitor: update hook version and start using the since token

Some file monitors like watchman will use a clock id to keep better track
of what changes happen in between calls to query the monitor.

Update the hook version handle the clock id so that it can be stored
with the index and the dirty flags.

Because the version of the fsmonitor hook will need to be updated to
support a clockid from watchman, create a config value for the hook
version that can be set.  This will help be able to stay backwards
compatible with version 1 of the hook but allow for a new version.

Signed-off-by: Kevin Willford <Kevin.Willford@microsoft.com>

kewillford committed:
fsmonitor: add fsmonitor test scripts for version 2

Need scripts used in testing fsmonitor using version 2.

Signed-off-by: Kevin Willford <Kevin.Willford@microsoft.com>

derrickstolee commented on fsmonitor.c:
```diff
@@ -32,17 +33,29 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
+		index += sizeof(uint64_t);
+	} else if (hdr_version == INDEX_EXTENSION_VERSION2) {
+		strbuf_addstr(&last_update, index);
+		index += last_update.len;
```
Do we need anything for a zero-terminated string?

derrickstolee commented on fsmonitor.c:
```diff
@@ -32,17 +33,29 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
-	istate->fsmonitor_last_update = get_be64(index);
-	index += sizeof(uint64_t);
+	istate->fsmonitor_last_update = strbuf_detach(&last_update, NULL);
+	strbuf_release(&last_update);
```
This `strbuf_release()` isn't needed if you just `detach`ed.

kewillford replied:
`strbuf_detach` does a `strbuf_init` so shouldn't it still need to be released?  Or is it because the `fsmonitor_last_update` has the memory that needs to be freed, release doesn't need to happen?
```
/**
 * Detach the string from the strbuf and returns it; you now own the
 * storage the string occupies and it is your responsibility from then on
 * to release it with `free(3)` when you are done with it.
 *
 * The strbuf that previously held the string is reset to `STRBUF_INIT` so
 * it can be reused after calling this function.
 */
char *strbuf_detach(struct strbuf *sb, size_t *sz)
{
	char *res;
	strbuf_grow(sb, 0);
	res = sb->buf;
	if (sz)
		*sz = sb->len;
	strbuf_init(sb, 0);
	return res;
}
```

derrickstolee commented on fsmonitor.c:
```diff
@@ -157,31 +169,38 @@ void refresh_fsmonitor(struct index_state *istate)
+	/* a fsmonitor process can return '/' to indicate all entries are invalid */
+	if (query_success && query_result.buf[bol] != '/') {
+		/* Mark all entries returned by the monitor as dirty */
+		for (;i < query_result.len; i++) {
```
hyper-nit: add a space after the first `;`

derrickstolee reviewed (commented):
This is looking promising!

kewillford committed:
fsmonitor: update documentation for hook version and watchman hooks

A new config value for core.fsmonitorHookVersion was added to be able
to force the version of the fsmonitor hook.

A version has been added to the fsmonitor hook to be able to handle
clock ids that are used with watchman.

Signed-off-by: Kevin Willford <Kevin.Willford@microsoft.com>

derrickstolee commented on t/t7519/fsmonitor-watchmanv2-debug:
```diff
@@ -0,0 +1,171 @@
+#!/usr/bin/perl
```
Do you still need this file?

derrickstolee reviewed (approved):
I don't really have any complaints. This looks like a good approach.

Have you started building installers so you can test this paired with the C# layer?

dscho commented on fsmonitor.c:
```diff
@@ -32,17 +33,27 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 	uint32_t ewah_size;
 	struct ewah_bitmap *fsmonitor_dirty;
 	int ret;
+	uint64_t tm;
```
I would like to see `tm` renamed to `tick` or `timestamp`.

jeffhostetler replied:
agreed.  `tm` is usually associated with `struct tm` and this usage would be confusing.

dscho commented on fsmonitor.c:
```diff
@@ -32,17 +33,27 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
+		strbuf_addstr(&last_update, index);
+		index += last_update.len + 1;
+	} else {
+		strbuf_release(&last_update);
```
No need to release that here, we have not allocated anything yet (`STRBUF_INIT` is a static initializer, it cannot allocate any memory).

kewillford replied:
For some reason I was thinking that every `STRBUF_INIT` needed a matching `strbuf_release` so there was a memory leak but if not all the better thanks!

dscho commented on fsmonitor.c:
```diff
@@ -218,18 +230,20 @@ void refresh_fsmonitor(struct index_state *istate)
-	/* Now that we've updated istate, save the last_update time */
-	istate->fsmonitor_last_update = last_update;
+	/* Now that we've updated istate, save the last_update_token */
+	istate->fsmonitor_last_update = strbuf_detach(&last_update_token, NULL);
```
Maybe we should `free(istate->fsmonitor_last_update)` just before this statement, just in case that we already took ownership of a token previously and now need to release its memory?

dscho commented on fsmonitor.c:
```diff
@@ -244,14 +258,16 @@ void add_fsmonitor(struct index_state *istate)
 		refresh_fsmonitor(istate);
 	}
+
+	strbuf_release(&last_update);
```
This is not necessary, as `strbuf_detach()` already re-sets `last_update` to no longer point to any allocated memory.

dscho commented on fsmonitor.c:
```diff
@@ -25,6 +26,18 @@ static void fsmonitor_ewah_callback(size_t pos, void *is)
+	if (!git_config_get_int("core.fsmonitorHookVersion", &hook_version)) {
+		if (hook_version != HOOK_INTERFACE_VERSION1 &&
+		    hook_version != HOOK_INTERFACE_VERSION2)
+			hook_version = -1;
```
We could write this as:

```c
int hook_version;

if (git_config_get_int("core.fsmonitorhookversion", &hook_version) ||
    (hook_version != HOOK_INTERFACE_VERSION1 &&
     hook_version != HOOK_INTERFACE_VERSION2))
        return -1;
return hook_version;
```

What do you think?

jeffhostetler replied:
Personally, I think I'd go for something like:

    int hook_version;
    
    if (git_config_get_int(..., &hook_version))
            return -1;
    if (hook_version == V1 ||
        hook_version == V2)
            return hook_version;
    warning(....); // maybe / maybe not
    return -1;

But that is just another opinion.

Johannes' version avoids the compile-time initialization, but also makes it fuzzy as to whether hook_version has been written to at the final return.   In both Kevin's and Johannes' version, the negation of the V1 and/or V2 clauses makes for a little more difficult reading.  IMHO.

dscho commented on fsmonitor.c:
```diff
@@ -173,27 +189,55 @@ void refresh_fsmonitor(struct index_state *istate)
+						break;
+					}
+				}	
+			}
```
How about adding this to handle failures:

```c
else if (hook_version < 0) {
        hook_version = HOOK_INTERFACE_VERSION1;
        if (!last_update_token.len)
                 		strbuf_addf(&last_update_token, "%"PRIu64"", last_update);
}
```

and also adding `if (hook_version < 0) hook_version = HOOK_INTERFACE_VERSION2;` to the block that handles success? That way, we could avoid sending an initial empty token to a v1 hook.

dscho commented on fsmonitor.c:
```diff
@@ -173,27 +189,55 @@ void refresh_fsmonitor(struct index_state *istate)
+						bol = i + 1;
+						break;
+					}
+				}	
```
We know that `query_result` is NUL-terminated, so we can use all the amenities provided by `string.h`. I propose to write it this way instead:

```c
strbuf_addstr(&last_update_token, query_result.buf);
if (last_update_token.len == query_result.buf)
    [... handle invalid payload, probably by ignoring it somehow...]
bol = last_update_token.len + 1;
```

kewillford replied:
Will you explain
`if (last_update_token.len == query_result.buf)`?
shouldn't it be 
`if (last_update_token.buf == query_result.buf)`

dscho replied:
Whoops. I think I meant to test for an empty token, i.e. `if (!last_update_token.len)`...

dscho commented on fsmonitor.c:
```diff
@@ -173,27 +189,55 @@ void refresh_fsmonitor(struct index_state *istate)
-		buf = query_result.buf;
-		bol = 0;
-		for (i = 0; i < query_result.len; i++) {
+		for (; i < query_result.len; i++) {
```
For v2, don't we start `i` at `bol - 1` by mistake?

In any case, I would prefer changing the code to `for (i = bol; [...])` here, i.e. keep the initial assignment of `i` strictly in this here block.

dscho commented on fsmonitor.c:
```diff
@@ -237,12 +281,15 @@ void refresh_fsmonitor(struct index_state *istate)
 void add_fsmonitor(struct index_state *istate)
 {
 	unsigned int i;
+	int hook_version;
```
Maybe we want to restrict the scope of this variable to the `if()` block below, i.e. move the declaration inside those curly braces?

dscho commented on fsmonitor.c:
```diff
@@ -222,6 +222,7 @@ void refresh_fsmonitor(struct index_state *istate)
 				istate->fsmonitor_last_update, &query_result);
 
 			if (query_success) {
+				strbuf_release(&last_update_token);
```
Better `strbuf_reset()`; This avoids releasing and immediately re-acquiring memory. It's a pattern to avoid.

dscho reviewed (commented):
I left a couple of comments/suggestions; Hopefully you find them useful!

Also, I'd like to see the `fixup!` "rebased away".

jeffhostetler commented on Documentation/githooks.txt:
```diff
@@ -490,9 +490,16 @@ fsmonitor-watchman
+nanoseconds since midnight, January 1, 1970.
+
+Version 2 takes two arguments, a version (2) and a token that is used
+to for identifying change since the token. For watchman this would be
```
typo: "... is used to for ..."

jeffhostetler commented on Documentation/githooks.txt:
```diff
@@ -490,9 +490,16 @@ fsmonitor-watchman
+Version 2 takes two arguments, a version (2) and a token that is used
+to for identifying change since the token. For watchman this would be
+a clock id. This version must output to stdout before the list of files
+the new token followed by a NUL.
```
nit:  maybe "This version must output to stdout the new token followed by a NUL before the list of files." ??

jeffhostetler commented on fsmonitor.c:
```diff
@@ -32,17 +46,27 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 
-	istate->fsmonitor_last_update = get_be64(index);
-	index += sizeof(uint64_t);
+	istate->fsmonitor_last_update = strbuf_detach(&last_update, NULL);
```
Just checking:  In the !v1 and !v2 else case, we never put anything in `last_update`.  Here we're stealing/detaching an empty buffer.  I assume that is OK ??

kewillford replied:
The `istate->fsmonitor_last_update` is used as a flag for checking if we are using fsmonitor in some places so we need to keep it something other than `0` for those checks.

dscho replied:
`strbuf_detach()` is always okay, even if nothing was added to the buffer: [`strbuf_detach()`](https://github.com/git/git/blob/53a06cf39b756eddfe4a2a34da93e3d04eb7b728/strbuf.c#L69-L78) calls [`strbuf_grow()`](https://github.com/git/git/blob/53a06cf39b756eddfe4a2a34da93e3d04eb7b728/strbuf.c#L90-L101) to make sure that _some_ memory is allocated.

Still, if we don't have to detach anything, it's better to avoid allocating something ;-)

jeffhostetler commented on fsmonitor.c:
```diff
@@ -32,17 +46,27 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
+		strbuf_addf(&last_update, "%"PRIu64"", tm);
+		index += sizeof(uint64_t);
+	} else if (hdr_version == INDEX_EXTENSION_VERSION2) {
+		strbuf_addstr(&last_update, index);
```
IIRC, `index` is now pointing at the `token` which is a null-terminated string, right?  There's an implicit assumption that the `last_update.len + 1` is at least as big as `sizeof(uint64_t)` because of the guard at the top.

kewillford replied:
Yeah I think that guard should be updated since the token could in theory be less than `sizeof(uint64_t)`.

jeffhostetler commented on fsmonitor.c:
```diff
@@ -32,17 +46,27 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
-	index += sizeof(uint64_t);
+	istate->fsmonitor_last_update = strbuf_detach(&last_update, NULL);
 
 	ewah_size = get_be32(index);
```
Just checking:  Previously, the header was a nicely aligned { ui32, ui64, ui32 } thing.  Now with the token, we have no guards on the size of the middle field, right?  So the `ewah_size` field might be filled using unaligned data, since `index` might not be a multiple of 4 at this point.  Again, just checking.

jeffhostetler commented on fsmonitor.c:
```diff
@@ -32,17 +46,27 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
-	index += sizeof(uint64_t);
+	istate->fsmonitor_last_update = strbuf_detach(&last_update, NULL);
 
 	ewah_size = get_be32(index);
```
Just checking:  Previously, the header was a nicely aligned { ui32, ui64, ui32 } thing.  Now with the token, we have no guards on the size of the middle field, right?  So the `ewah_size` field might be filled using unaligned data, since `index` might not be a multiple of 4 at this point.  Again, just checking.

kewillford replied:
I was wondering about this.  Does the ewah bitmap need to be aligned?  If so I will need to do some work to make sure it is.

jeffhostetler replied:
not sure about ewah's themselves, you might ask @derrickstolee about that (he's played with them and I haven't).  I was mainly worried about an unaligned processor fault inside get_be32() on finicky processors.

dscho replied:
`get_be32()` is defined in two versions, one guarded by the unwieldy condition:

```c
#if !defined(NO_UNALIGNED_LOADS) && ( \
    defined(__i386__) || defined(__x86_64__) || \
    defined(_M_IX86) || defined(_M_X64) || \
    defined(__ppc__) || defined(__ppc64__) || \
    defined(__powerpc__) || defined(__powerpc64__) || \
    defined(__s390__) || defined(__s390x__))
```

So I think we're actually safe, we do not have to worry that the `get_be*()` function family might stumble over alignment issues: they don't.

dscho replied:
(see `compat/bswap.h`)

jeffhostetler replied:
thanks!

kewillford commented on fsmonitor.c:
```diff
@@ -32,17 +46,27 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
+	uint64_t tm;
+	struct strbuf last_update = STRBUF_INIT;
 
 	if (sz < sizeof(uint32_t) + sizeof(uint64_t) + sizeof(uint32_t))
```
This check is wrong now.

jeffhostetler commented on fsmonitor.c:
```diff
@@ -162,26 +189,56 @@ void refresh_fsmonitor(struct index_state *istate)
 	 * should be inclusive to ensure we don't miss potential changes.
 	 */
 	last_update = getnanotime();
```
Just a thought:  `getnanotime()` has a bunch of slop in it.  For example, the first call to it within the process always gets a value ending with `000` on Windows (observed when I started trace2).  And there's some other mysterious magic in there that I don't completely understand.

We might want to consider just getting the whole seconds since epoch, subtract 1 or 2, and then convert that to nanoseconds for storage in the header.  And err on the side of caution.

derrickstolee commented:
I'm thrilled that this is in v2.26.0-rc0, so will be in our fork as part of vfs-2.26.0. When I get installers working for #251, then we can test microsoft/scalar#255 with that build.
